File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // freq vector for storing value of char or it's like a mark
        vector<int> umap(256,-1);
        int n = s.size();
        int l = 0, r = 0;   // l-left, r-right pointer
        int len = 0;    // initial lenr
        while(r < n){
            // if char is present in freq then we are gonna to update l-left by which idx it's appear
            if(umap[s[r]] != -1) l = max(umap[s[r]] + 1,l);
            // updating freq 
            umap[s[r]] = r;
            // updating len
            len = max(len, r - l + 1);  // we are doing r-l+1 bcaz have to return substring size and idx.
            r++;
        }
        return len;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int freq[] = new int[256];
        Arrays.fill(freq, 0);
        int l = 0, r = 0, n = s.length();
        int ans = 0;
        while(r < n){
            freq[s.charAt(r)]++;
            while(freq[s.charAt(r)] > 1){
                freq[s.charAt(l)]--;
                l++;
            }
            int val = r - l + 1;
            if(val > ans) ans = val;
            r++;
        }
        return ans;
    }
}